ast_tools/stack.py

```diff
@@ -0,0 +1,89 @@
+'''
+Functions and classes the inspect or modify the stack
+'''
+
+import inspect
+import typing as tp
+import types
+import functools
+import itertools
+
+from collections import ChainMap
+from contextlib import contextmanager
+import logging
+
+_SKIP_FRAME_DEBUG_NAME = '_AST_TOOLS_STACK_DEBUG_SKIPPED_FRAME_'
+_SKIP_FRAME_DEBUG_VALUE = 0xdeadbeaf
+_SKIP_FRAME_DEBUG_STMT = f'{_SKIP_FRAME_DEBUG_NAME} = {_SKIP_FRAME_DEBUG_VALUE}'
+_SKIP_FRAME_DEBUG_FAIL = False
+
+class SymbolTable(tp.NamedTuple):
+    locals: tp.Mapping[str, tp.Any]
+    globals: tp.Dict[str, tp.Any]
+
+def get_symbol_table(
+        decorators: tp.Optional[tp.Sequence[inspect.FrameInfo]] = None
+        ) -> SymbolTable:
+    exec(_SKIP_FRAME_DEBUG_STMT)
+    locals = ChainMap()
+    globals = ChainMap()
+
+    if decorators is None:
+        decorators = set()
+    else:
+        decorators = {f.__code__ for f in decorators}
+    decorators.add(get_symbol_table.__code__)
+
+
+    stack = inspect.stack()
+    for i in range(len(stack) - 1, 0, -1):
+        frame = stack[i]
+        if frame.frame.f_code in decorators:
+            continue
+        debug_check = frame.frame.f_locals.get(_SKIP_FRAME_DEBUG_NAME, None)
+        if debug_check == _SKIP_FRAME_DEBUG_VALUE:
+            if _SKIP_FRAME_DEBUG_FAIL:
+                raise RuntimeError(f'{frame.function} @ {frame.filename}:{frame.lineno} might be leaking names')
+            else:
+                logging.debug(f'{frame.function} @ {frame.filename}:{frame.lineno} might be leaking names')
+        locals = locals.new_child(stack[i].frame.f_locals)
+        globals = globals.new_child(stack[i].frame.f_globals)
+    return SymbolTable(locals=locals, globals=globals)
+
+def inspect_symbol_table(
+        fn: tp.Callable, # tp.Callable[[SymbolTable, ...], tp.Any],
+        *,
```

Comment:
Is the above commented type signature needed? Is it just a more "specific" type annotation that's not actually supported by Callable? If so, fine to leave it.

Reply:
Also, not exactly sure what this `*` is doing, is it capturing any following arguments but not using them (hence the lack of name)? What's the purposes of allowing these arguments but not using them?

Reply:
As you suspected the comment is just a more specific type annotation that's not actually supported by Callable.

The * indicates keyword only arguments.  https://www.python.org/dev/peps/pep-3102/
I really don't want people accidentally setting decorators because it can produce entirely inscrutable errors.

Reply:
Got it, learned something new today!